pkg/controllers/user/windows/linux_node_taints.go

```diff
@@ -0,0 +1,62 @@
+package windows
+
+import (
+	"fmt"
+
+	"github.com/rancher/rancher/pkg/taints"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var (
+	nodeTaint = v1.Taint{
+		Key:    "cattle.io/os",
+		Value:  "linux",
+		Effect: v1.TaintEffectNoSchedule,
+	}
+)
+
+// NodeTaintsController This controller will only run on the cluster with windowsPreferred is true.
+// It will add taints to the nodes with label beta.kubernetes.io/os=linux.
+type NodeTaintsController struct {
+	nodeClient v3.NodeInterface
+}
+
+func (n *NodeTaintsController) sync(key string, obj *v3.Node) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return obj, nil
+	}
+	found := false
+	for _, hostOSLabel := range HostOSLabels {
+		if hostOSLabel.AsSelector().Matches(labels.Set(obj.Status.NodeLabels)) {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return obj, nil
+	}
+
+	// NodeTaints is updating, skip this sync
+	if obj.Spec.UpdateTaintsFromAPI != nil {
+		return obj, nil
+	}
+
+	taintSet := taints.GetTaintSet(obj.Spec.InternalNodeSpec.Taints)
+	// taint exists on nodes
+	if _, ok := taintSet[taints.GetTaintsString(nodeTaint)]; ok {
+		return obj, nil
+	}
+
+	newObj := obj.DeepCopy()
+	newObj.Spec.DesiredNodeTaints = append(newObj.Spec.DesiredNodeTaints, newObj.Spec.InternalNodeSpec.Taints...)
```

Comment:
q: what is this for? I thought we'd only be adding `nodeTaint` here

Reply:
The desired node taints field should be nil if the `obj.Spec.UpdateTaintsFromAPI` is nil. That is the normal state for node. It means that no need to update node taints.
When we want to update the taints, we will set both `DesiredNodeTaints` and `UpdateTaintsFromAPI`. Because we need to support the removing node taints case, we need to set the whole expected node taint set into `DesiredNodeTaints`.
That is the reason why I need to copy `InternalNodeSpec.Taints` to desired node taints and append our taint into that list.